src/apis/Event.re

```diff
@@ -1,106 +1,143 @@
+  [@bs.get] external nativeEvent: T.target('a) => 'a = "nativeEvent";
+  [@bs.send] external persist: T.target('a) => unit = "persist";
+  [@bs.get] [@bs.return nullable]
+  external target: T.target('a) => option(float) = "target";
+  [@bs.get] external timeStamp: T.target('a) => float = "timeStamp";
+  [@bs.get] [@bs.return nullable]
+  external _type: T.target('a) => option(string) = "type";
 };
 
-type responderSyntheticEvent('a) = {
-  .
-  // synthethicEvent keys
-  "bubbles": Js.Nullable.t(bool),
-  "cancelable": Js.Nullable.t(bool),
-  "currentTarget": float,
-  "defaultPrevented": Js.Nullable.t(bool),
-  "dispatchConfig": {. "registrationName": string},
-  "eventPhase": Js.Nullable.t(float),
-  [@bs.meth] "preventDefault": unit => unit,
-  [@bs.meth] "isDefaultPrevented": unit => bool,
-  [@bs.meth] "stopPropagation": unit => unit,
-  [@bs.meth] "isPropagationStopped": unit => bool,
-  "isTrusted": Js.Nullable.t(bool),
-  "nativeEvent": 'a,
-  [@bs.meth] "persist": unit => unit,
-  "target": Js.Nullable.t(float),
-  "timeStamp": float,
-  "_type": Js.Nullable.t(string),
-  // responderSyntheticEvent additional key
-  "touchHistory": {
-    .
-    "indexOfSingleActiveTouch": float,
-    "mostRecentTimeStamp": float,
-    "numberActiveTouches": float,
-    "touchBank":
-      array({
-        .
-        "touchActive": bool,
-        "startPageX": float,
-        "startPageY": float,
-        "startTimeStamp": float,
-        "currentPageX": float,
-        "currentPageY": float,
-        "currentTimeStamp": float,
-        "previousPageX": float,
-        "previousPageY": float,
-        "previousTimeStamp": float,
-      }),
-  },
+module SyntheticEvent = {
+  type t_('a) = {
+    bubbles: Js.Nullable.t(bool),
+    cancelable: Js.Nullable.t(bool),
+    currentTarget: float,
+    defaultPrevented: Js.Nullable.t(bool),
+    dispatchConfig: registrationName,
+    eventPhase: Js.Nullable.t(float),
+    isTrusted: Js.Nullable.t(bool),
+    nativeEvent: 'a,
+    target: Js.Nullable.t(float),
+    timeStamp: float,
+    [@bs.as "type"]
+    _type: Js.Nullable.t(string),
+  }
+  and registrationName = {registrationName: string};
+
+  include EventMethods({
+    type target('a) = t_('a);
+    type r = registrationName;
+  });
+};
+
+module ResponderSyntheticEvent = {
+  type t_('a) = {
+    bubbles: Js.Nullable.t(bool),
+    cancelable: Js.Nullable.t(bool),
+    currentTarget: float,
+    defaultPrevented: Js.Nullable.t(bool),
+    dispatchConfig: registrationName,
+    eventPhase: Js.Nullable.t(float),
+    isTrusted: Js.Nullable.t(bool),
+    nativeEvent: 'a,
+    target: Js.Nullable.t(float),
+    timeStamp: float,
+    [@bs.as "type"]
+    _type: Js.Nullable.t(string),
+    touchHistory,
+  }
+  and registrationName = {registrationName: string}
+  and touchHistory = {
+    indexOfSingleActiveTouch: float,
+    mostRecentTimeStamp: float,
+    numberActiveTouches: float,
+    touchBank: array(touchBank),
+  }
+  and touchBank = {
```

Comment:
No dependent definition necessary.

Reply:
Same here, organisational choice. I don't suppose there's a significant performance impact. I expect there can't be further impact once `reason-react-native` is compiled.

Reply:
Actually gentype cannot handle this. Not sure if that’s relevant but I also believe it’s better to keep types as simple as possible so it’s easier for the compiler.

Reply:
I agree with @jfrolich, and I think it would also make it easier for the reader, not just the compiler. When I read "let...and", at first glance I automatically assume that this is a recursive definition. Then I start scanning for where the recursion actually is and find none.